Equations of state

nilsdeppe opened the pull request:
## Proposed changes

- Add support for checking member functions with python
- Add polytropic EoS
- Add ideal fluid EoS
- Add dark energy EoS

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
Add polytropic fluid equation of state

nilsdeppe committed:
Add member function support to CheckWithRandomValues

nilsdeppe committed:
Add ideal fluid equation of state

nilsdeppe committed:
Add dark energy fluid equation of state

fmahebert commented on src/PointwiseFunctions/EquationsOfState/PolytropicFluid.hpp:
```diff
@@ -0,0 +1,111 @@
+      "A polytropic fluid equation of state.\n"
+      "The pressure is related to the rest mass density by p = K rho ^ Gamma, "
+      "where p is the pressure, rho is the rest mass density, K is the "
+      "polytropic  constant, and Gamma is the polytropic exponent.The "
```
Also remove the extra whitespace between `polytropic  constant`.

nilsdeppe replied:
space before `The`

fmahebert commented on src/PointwiseFunctions/EquationsOfState/PolytropicFluid.hpp:
```diff
@@ -0,0 +1,111 @@
+ *
+ * A polytropic equation of state \f$p=K\rho^{\Gamma}\f$ where \f$K\f$ is the
+ * polytropic constant and \f$\Gamma\f$ is the polytropic exponent. The
+ * polytropic exponent is related to the polytropic index by
```
"the polytropic index \f$N_p\f$ by"

kidder commented on tests/Unit/Pypp/CheckWithRandomValues.hpp:
```diff
@@ -281,4 +282,288 @@ void check_with_random_values(
+                                    make_not_null(&generator),
+                                    make_not_null(&(distributions[0]))),
+      '0')...};
+  (klass.*f)(make_not_null(&std::get<ResultIs>(results))...,
```
try using a lambda

kidder commented on cmake/SpectreParseTests.py:
```diff
@@ -14,6 +14,7 @@
                 "DataStructures",
                 "Domain",
                 "ErrorHandling",
+                "EquationsOfState",
```
RIP

kidder commented on docs/GroupDefs.hpp:
```diff
@@ -105,6 +105,11 @@
  * Macros and functions used for handling errors
  */
 
+/*!
```
RIP again

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,207 @@
+}  // namespace EquationsOfState
+/// \endcond
+
+/// Contains all equations of state, including base classes
```
class

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,207 @@
+/*!
+ * \ingroup EquationsOfStateGroup
+ * \brief Macro used to generate forward declarations of member functions in
+ * base classes
```
derived

nilsdeppe commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,207 @@
+  // @}
+
+  // @{
+  /// Computes \f$\kappa p/\rho^2=(p/\rho^2)\partial p / \partial \epsilon\f$
```
define `p`

nilsdeppe commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,207 @@
+      const = 0;
+
+  // @{
+  /// Computes the pressure from the rest mass density \f$\rho\f$ and the
```
`pressure \f$p\f$` liberally (and in `chi`)

nilsdeppe commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,207 @@
+  // @}
+
+  // @{
+  /// Computes the sound speed squared, \f$c_s{}^2\f$, from the rest mass
```
parens around `c_s`

kidder commented on src/PointwiseFunctions/EquationsOfState/PolytropicFluid.cpp:
```diff
@@ -0,0 +1,219 @@
+    : polytropic_constant_(polytropic_constant),
+      polytropic_exponent_(polytropic_exponent) {}
+
+template <bool IsRelativistic>
```
macro the calls to impls

kidder commented on tests/Unit/PointwiseFunctions/EquationsOfState/Test_PolytropicFluid.cpp:
```diff
@@ -0,0 +1,149 @@
+        &EoS::EquationOfState<IsRelativistic>::pressure_from_internal_energy};
+    pypp::check_with_random_values<2>(
+        func, *eos, "TestFunctions", "polytropic_pressure",
+        {{{1.0e-4, 4.0}, {-10.0, 10.0}}},
```
h  > 1 for Relativistic, > 0 for Newtonian ; epsilon > 0

nilsdeppe commented on src/PointwiseFunctions/EquationsOfState/IdealFluid.hpp:
```diff
@@ -0,0 +1,111 @@
+ * p = \rho \epsilon(\gamma-1)
+ * \f]
+ * where \f$\rho\f$ is the rest mass density, \f$\epsilon\f$ is the specific
+ * internal energy, and \f$\gamma\f$ is the so-called adiabatic index.
```
remove `so-called`

kidder commented on src/PointwiseFunctions/EquationsOfState/IdealFluid.cpp:
```diff
@@ -0,0 +1,232 @@
+    const Scalar<DataType>& /*rest_mass_density*/,
+    const Scalar<DataType>& specific_internal_energy) const noexcept {
+  const double temp = adiabatic_index_ - 1.0;
+  return Scalar<DataType>{square(temp) * get(specific_internal_energy)};
```
inline

kidder commented on src/PointwiseFunctions/EquationsOfState/IdealFluid.hpp:
```diff
@@ -0,0 +1,111 @@
+      const Scalar<DataType>& rest_mass_density,
+      const Scalar<DataType>& specific_internal_energy) const noexcept;
+
+  template <class DataType>
```
macro the impl declarations

kidder commented on tests/Unit/PointwiseFunctions/EquationsOfState/Test_IdealFluid.cpp:
```diff
@@ -0,0 +1,138 @@
+        std::make_tuple(
+            make_with_value<Scalar<T>>(used_for_size, adiabatic_index)),
+        used_for_size);
+  };
```
can there be one test_eos function?

fmahebert commented on src/PointwiseFunctions/EquationsOfState/DarkEnergyFluid.hpp:
```diff
@@ -0,0 +1,116 @@
+ * \f$z\f$.
+ *
+ * \warning The numerical implementation requires \f$w\neq-1\f$ and
+ * \f$w\neq0\f$.
```
Maybe I missed it, but couldn't this be enforced in the code?

fmahebert commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,215 @@
+
+  // @{
+  /// Computes the pressure from the rest mass density \f$\rho\f$ and the
+  /// specific internal energy \f$\epsilon\f$
```
Personally, I would add a period at end of this sentence (and similar cases below)...

mar-celine commented on src/PointwiseFunctions/EquationsOfState/PolytropicFluid.cpp:
```diff
@@ -0,0 +1,219 @@
+Scalar<double> PolytropicFluid<IsRelativistic>::specific_enthalpy_from_pressure(
+    const Scalar<double>& rest_mass_density,
+    const Scalar<double>& pressure) const noexcept {
+  return specific_ethalpy_from_pressure_impl(rest_mass_density, pressure);
```
`ethalpy` (5 occurrences)

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,343 @@
+  // @{
+  /// Computes the pressure \f$p\f$ from the rest mass density \f$\rho\f$
+  virtual Scalar<double> pressure_from_density(
+      const Scalar<decltype((void)Is, double{0.0})>&...) const noexcept = 0;
```
hide this crap in a metafunction/type alias :)

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,343 @@
+  // @}
+
+  // @{
+  /// Computes \f$\chi=\partial p / \partial \rho\f$ from the rest mass
```
define p

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,343 @@
+  // @}
+
+  // @{
+  /// Computes \f$\kappa p/\rho^2=(p/\rho^2)\partial p / \partial \epsilon\f$
```
define p

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,343 @@
+    return chi_from_density_impl(rest_mass_density BOOST_PP_REPEAT(            \
+        BOOST_PP_SUB(DIM, 1), EQUATION_OF_STATE_FORWARD_ARGUMENTS, UNUSED));   \
+  }                                                                            \
+  TEMPLATE                                                                     \
```
simplify (with another macro?) to just pass the function name

kidder commented on src/PointwiseFunctions/EquationsOfState/PolytropicFluid.hpp:
```diff
@@ -0,0 +1,96 @@
+template <bool IsRelativistic>
+class PolytropicFluid : public EquationOfState<IsRelativistic, 1> {
+ public:
+  static constexpr bool is_relativistic = IsRelativistic;
```
these should be inherited from the base class

kidder commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,343 @@
+  ~EquationOfState() override = default;
+
+  WRAPPED_PUPable_abstract(EquationOfState);  // NOLINT
+  virtual std::unique_ptr<EquationOfState<IsRelativistic, ThermodynamicDim>>
```
let's get rid of the get_clone (don't want to clone tables!)

kidder commented on tests/Unit/PointwiseFunctions/EquationsOfState/TestHelpers.hpp:
```diff
@@ -0,0 +1,190 @@
+  // Is does not go full range because we want to be able to go from enthalpy to
+  // rest mass density, which means we need to be able to adjust the first
+  // bounds independently
+  const std::array<std::pair<double, double>, 2> random_value_bounds{
```
comment what variables are being bounded

kidder commented on tests/Unit/PointwiseFunctions/EquationsOfState/TestHelpers.hpp:
```diff
@@ -0,0 +1,190 @@
+       " equation of state"s)
+
+  static_assert(sizeof...(Is) == ThermodynamicDim - 1, "");
+  // Is does not go full range because we want to be able to go from enthalpy to
```
move comment above static_assert and INFO below

kidder commented on tests/Unit/PointwiseFunctions/EquationsOfState/TestHelpers.hpp:
```diff
@@ -0,0 +1,190 @@
+            make_with_value<Scalar<T>>(used_for_size, member_args)...),
+        used_for_size);
+    make_overloader(
+        [&](const std::integral_constant<size_t, 1>& /*meta*/,
```
meta -> thermodynamic_dim

nilsdeppe committed:
Add identity_t type alias for turning int packs into types

nilsdeppe committed:
Add more IWYU maps for boost

nilsdeppe committed:
Update clang-tidy checks for clang-tidy-6.0

nilsdeppe committed:
Add gsl::owner

mar-celine commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -1418,4 +1418,12 @@ struct function_info_impl<Ret (Class::*)(Args...) volatile> {
+/*!
+ * \ingroup TypeTraitsGroup
+ * \brief Given a type `T` and possibly a `size_t` evaluates to `T`. Useful for
+ * turning a `std::index_sequence` into a pack expansion of types.
```
point to an example using parameter pack expansion

kidder commented:
@wthrowe we think this is ready

wthrowe commented on src/PointwiseFunctions/EquationsOfState/EquationOfState.hpp:
```diff
@@ -0,0 +1,322 @@
+
+/*!
+ * \ingroup EquationsOfStateGroup
+ * \brief Base class for equations of state depending on
```
depending on...?

wthrowe commented on tests/Unit/PointwiseFunctions/EquationsOfState/Test_IdealFluid.cpp:
```diff
@@ -0,0 +1,60 @@
+  TestHelpers::EquationsOfState::check(EoS::IdealFluid<true>{5.0 / 3.0},
+                                       "ideal_fluid", d_for_size, 5.0 / 3.0);
+  TestHelpers::EquationsOfState::check(EoS::IdealFluid<true>{4.0 / 3.0},
+                                       "ideal_fluid", dv_for_size, 4.0 / 3.0);
```
This is two checks repeated twice each.

wthrowe commented on tests/Unit/PointwiseFunctions/EquationsOfState/Test_DarkEnergyFluid.cpp:
```diff
@@ -0,0 +1,63 @@
+  TestHelpers::EquationsOfState::check(EoS::DarkEnergyFluid<true>(1.0),
+                                       "dark_energy_fluid", dv_for_size, 1.0);
+  TestHelpers::EquationsOfState::check(EoS::DarkEnergyFluid<true>(-1.0),
+                                       "dark_energy_fluid", dv_for_size, -1.0);
```
Is there a logic to the order of these checks?

nilsdeppe replied:
nope

wthrowe commented on tests/Unit/PointwiseFunctions/EquationsOfState/Test_DarkEnergyFluid.cpp:
```diff
@@ -0,0 +1,63 @@
+      test_factory_creation<EoS::EquationOfState<true, 2>>(
+          {"  DarkEnergyFluid:\n"
+           "    ParameterW: 0.3333333333333333\n"}),
+      "dark_energy_fluid", dv_for_size, 1.0 / 3.0);
```
This is two copies of the same two checks.

wthrowe commented on tests/Unit/Utilities/Test_TypeTraits.cpp:
```diff
@@ -1076,4 +1078,18 @@ struct TestFunctions {
+/// [example_identity_t]
+template <size_t... Is>
+struct IdentityPackExample<std::index_sequence<Is...>> {
+  using type = IdentityPackExample<tt::identity_t<double, Is>...>;
```
The recursive nature of this example is confusing, particularly because the declaration of the struct is not included in the example so it is not clear that it is variadic.  How about just using a tuple or something like that?

wthrowe commented on src/Utilities/Gsl.hpp:
```diff
@@ -131,6 +131,13 @@ SPECTRE_ALWAYS_INLINE constexpr const T& at(std::initializer_list<T> cont,
+ * \brief Mark a raw pointer as owning its data
+ */
+template <class T, Requires<std::is_pointer<T>::value> = nullptr>
+using owner = T;
```
This is unused.  I think we should drop it because we want to discourage creating owning raw pointers.

nilsdeppe replied:
clang-tidy 6.0 requests this (using gsl::owner when an owning raw pointer is used), so we certainly should not get rid of it. I think we should handle not wanting owning raw pointers in one of two ways:
- Have a static_assert be triggered if `owner` is used so that when clang-tidy complains, someone uses gsl::owner and hits a static_assert (plus documentation for owner)
- We only add documentation discouraging its use.

I'm fine with either, but since we really won't ever have owning raw pointers the first is better.

I noticed we were missing owner because of clang-tidy complaining inside the pup function for unique_ptr.

kidder reviewed (commented):
squash

nilsdeppe committed:
Disable cppcheck builds

cppcheck can't parse C++14 correctly so we can't use it. Also, everything it has
suggested so far has been a false positive.
